ehrapy/tools/cohort_tracking/_cohort_tracker.py

```diff
@@ -0,0 +1,332 @@
+import copy
+from typing import Any, Union
+
+import graphviz
+import matplotlib.pyplot as plt
+import numpy as np
+import pandas as pd
+import seaborn as sns
+from scanpy import AnnData
+from tableone import TableOne
+
+
+def _check_columns_exist(df, columns):
+    if not all(col in df.columns for col in columns):
+        missing_columns = [col for col in columns if col not in df.columns]
+        raise ValueError(f"Columns {missing_columns} not found in dataframe.")
+
+
+# from tableone: https://github.com/tompollard/tableone/blob/bfd6fbaa4ed3e9f59e1a75191c6296a2a80ccc64/tableone/tableone.py#L555
+def _detect_categorical_columns(data) -> list:
+    # assume all non-numerical and date columns are categorical
+    numeric_cols = set(data._get_numeric_data().columns.values)
+    date_cols = set(data.select_dtypes(include=[np.datetime64]).columns)
+    likely_cat = set(data.columns) - numeric_cols
+    # mypy absolutely looses it if likely_cat is overwritten to be a list
+    likely_cat_no_dates = list(likely_cat - date_cols)
+
+    # check proportion of unique values if numerical
+    for var in data._get_numeric_data().columns:
+        likely_flag = 1.0 * data[var].nunique() / data[var].count() < 0.005
+        if likely_flag:
+            likely_cat_no_dates.append(var)
+    return likely_cat_no_dates
+
+
+class CohortTracker:
+    def __init__(self, adata: AnnData | pd.DataFrame, columns: list = None, categorical: list = None, *args: Any):
+        """Track cohort changes over multiple filtering or processing steps.
+
+        This class offers functionality to track and plot cohort changes over multiple filtering or processing steps,
+        enabling the user to monitor the impact of each step on the cohort.
+
+        Tightly interacting with the `tableone` package [1].
+        Args:
+            adata: :class:`~anndata.AnnData` or :class:`~pandas.DataFrame` object to track.
+            columns: List of columns to track. If `None`, all columns will be tracked.
+            categorical: List of columns that contain categorical variables, if not given will be inferred from the data.
+
+        References
+        ----------
+        [1] Tom Pollard, Alistair E.W. Johnson, Jesse D. Raffa, Roger G. Mark; tableone: An open source Python package for producing summary statistics for research papers, Journal of the American Medical Informatics Association, Volume 24, Issue 2, 1 March 2017, Pages 267–271, https://doi.org/10.1093/jamia/ocw117
+
+        """
+        if isinstance(adata, AnnData):
+            df = adata.obs
+        elif isinstance(adata, pd.DataFrame):
+            df = adata
+        else:
+            raise ValueError("adata must be an AnnData or a DataFrame.")
```

Comment:
Yeah the name `adata` is a bit tricky here. We could either name it `data` or drop DataFrame support. Or just keep the name `adata`

Reply:
solved by not allowing for df